[자동차 경주 구현] 팡태 미션 제출합니다.

Hevton opened the pull request:
리뷰어님 안녕하세요! 저는 안드로이드 6기 팡태(황태준)입니다!
시간 내어 코드를 읽어 주셔서 정말 감사드립니다.

미션을 진행하면서 몇가지 여쭤보고 싶은 점이 생겼는데, 혹시 답변 남겨주실 수 있으시다면 정말 감사드리겠습니다.

1. 클래스 분리가 적절한지 피드백을 받고 싶습니다.
MVC 모델 기반으로 진행하고자 했습니다. 다양한 도메인을 구현했는데, 시도 횟수를 입력받는 과정까지 도메인 클래스로 구현하여 테스트 코드를 작성하는데 이점을 누리고자 했는데, 이 부분에 대한 의견을 구하고 싶습니다. 또한 이외에도 클래스 분리가 전체적으로 합리적인지 궁금합니다.

2. 데이터 클래스를 이용한 경우, 클래스를 이용한 경우
페어와 함께 이번 미션을 진행하면서 클래스 생성에 두 가지 방법을 도입했습니다.
A. Car 클래스는 data 클래스로 정의하고, 유효성 검증을 init에서 진행했습니다.
B. 그리고 이와달리 TryCount와 Cars는 일반 클래스로 정의한 뒤에 팩토리 메서드를 통해 유효성 검증과 객체 생성을 진행했습니다.

데이터 클래스는 DTO처럼 일반적으로 데이터를 주고받는 과정에서만 사용하고, 함수 구현이나 유효성 검증 등을 진행하지 않아야 하나 싶어 일반 클래스로 구현하고 팩토리 메서드를 통해 객체를 생성함과 동시에 유효성 검증을 진행하는 방향(B)을 더 바라봐야 할지, 아니면 A 방향도 고려할 수 있을지 궁금합니다.

3. 유효성 검증 방법
이번 미션에서 도메인의 유효성 검증을 진행할 때, 확장함수를 사용하였고, 해당 확장함수는 관련성이 있는 도메인 클래스 파일 내에 넣어주었습니다. 이러한 유효성 검증 방식과 분리가 현재 제 능력의 상한선이라고 생각되기에, 이에 대한 리뷰어님의 시야와 의견이 궁금합니다.

읽어주셔서 감사드립니다.

sh1mj1 committed:
docs(README): 기능 명세서 작성

sh1mj1 committed:
build(deps): 여러 인자 테스트 위한 의존성 추가

sh1mj1 committed:
feat(Car): 자동차 이름에 대한 유효성 검증

- 공백을 허용하지 않는다
- 5자 이하여야 한다
관련하여 단위 테스트 완료

sh1mj1 committed:
feat(RandomNumberGenerator): 0 부터 9 까지의 무작위 숫자를 생성하는 숫자 생성기 구현

- 이 클래스는 NumberGenerator 를 구현한다

sh1mj1 committed:
docs(README): NumberGenerator 구현체 방식 변경

- 이전: 테스트용 Generator로 Move/Stop 두 클래스를 사용
- 이후: 테스트 도중 원하는 숫자를 직접 지정할 수 있게끔 하나로 통합

sh1mj1 committed:
feat(Car): 자동차가 움직이는 기능 구현

- 0 부터 3 이면 움직이지 않는다.
- 4 부터 9 이면 움직인다.
- ExplicitNumberGenerator 를 이용하여 직접 지정한 숫자로 단위 테스트

sh1mj1 committed:
feat(TryCount): 시도 횟수를 생성하는 팩토리 메서드와 유효성 검증 구현

- 시도 횟수는 정수여야 합니다
- 시도 횟수는 0보다 큰 수여야 합니다
관련하여 단위 테스트 완료

sh1mj1 committed:
feat(InputView): 경주할 자동차 이름을 입력받는다

sh1mj1 committed:
feat(InputView): 시도할 횟수를 입력받는다

sh1mj1 committed:
docs(README): 게임에 출전할 모든 자동차를 담는 클래스 요구사항을 추가

- 이전: RacingGame 클래스에서 모든 자동차를 가지며 유효성 검증 수행
- 이후: 모든 자동차를 가지는 클래스(Cars) 별도 생성. 여기서 유효성 검증 수행

sh1mj1 committed:
feat(Cars): 게임에 출전하는 모든 차 입력 처리

- 자동차 이름에 중복이 없어야 한다
관련하여 단위 테스트 완료

sh1mj1 committed:
feat(OutputView): 경주 결과를 출력한다

- 각 실행(phase)마다 자동차 위치를 출력한다
- 최종 우승자를 출력한다

sh1mj1 committed:
docs(README): Cars 에 경주 한 턴(phase)을 실행하는 요구사항 추가

sh1mj1 committed:
feat(Cars): Cars 에 경주 한 턴(phase)을 실행하는 기능 구현

- 전진, 멈춤에 대한 단위 테스트 구현

sh1mj1 committed:
fix(OutputView): 우승자 출력 형식 변경

sh1mj1 committed:
feat(RacingGame): Cars 을 시도 횟수 만큼 전진/멈춤 시키는 경주를 진행한다.

- 출전할 자동차들과 시도 횟수를 입력받아서 경주 진행 및 출력

sh1mj1 committed:
refactor(RandomNumberGenerator): 랜덤 수를 생성하는 클래스를 싱클톤으로 변경

- 여러 객체를 생성할 필요가 없음

sh1mj1 committed:
style(gitkeep): 오직 패키지 구조를 유지하기 존재했던 파일 제거

sh1mj1 committed:
style(constant): 문자열을 상수화

Gyuil-Hwnag commented on README.md:
```diff
@@ -1 +1,39 @@
-# kotlin-racingcar
\ No newline at end of file
+# kotlin-racingcar
+
+## 기능 명세
```
README 작성 좋습니다👍

Hevton replied:
감사합니다 ㅎㅎ

Gyuil-Hwnag commented on src/main/kotlin/controller/RacingGame.kt:
```diff
@@ -0,0 +1,35 @@
+    private val numberGenerator: NumberGenerator
+) {
+
+    fun start() {
+        val cars = readCars()
+        val tryCount = readTryCount()
+
+        race(cars, tryCount)
+    }
+
+    private fun readCars(): Cars = Cars.from(inputView.readCars(), numberGenerator)
+
+    private fun readTryCount(): TryCount = TryCount.from(inputView.readTryCount())
+
+    private fun race(cars: Cars, tryCount: TryCount) {
+        outputView.printResultHeader()
+        repeat(tryCount.count) {
+            cars.startPhase()
+            outputView.printPhase(cars)
+        }
+        outputView.printWinner(cars)
+    }
```
각각의 기능별로 함수 구분 너무 잘해주셨네요💯

Hevton replied:
감사합니다!

Gyuil-Hwnag commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -0,0 +1,24 @@
+
+    private fun canMove(): Boolean = numberGenerator.generate() >= MOVE_LOWER_BOUND
+
+    companion object {
+        private const val MOVE_LOWER_BOUND = 4
+        private const val EXCEPTION_CAR_NAME_IS_BLANK = "자동차 이름은 공백을 허용하지 않습니다"
+        private const val EXCEPTION_INVALID_CAR_NAME_LENGTH = "자동차 이름은 다섯글자 이하여야 합니다"
+    }
```
companion object를 활용한 상수 관리 좋습니다💯

Gyuil-Hwnag commented on src/main/kotlin/domain/Cars.kt:
```diff
@@ -0,0 +1,25 @@
+    }
+}
+
+fun String.validateCars(): List<String> {
+    val carNames = this.split(",")
+    require(carNames.size == carNames.toSet().size) { EXCEPTION_DUPLICATED_NAME }
+    return carNames
+}
+
+const val EXCEPTION_DUPLICATED_NAME = "중복된 이름이 존재합니다."
```
fun validateCars() 의 경우 Cars 클래스에서만 사용하는 함수라면 Cars 클래스 내부에 선언해도 좋을 것 같다는 생각이 듭니다!!
추후 다른 기능들을 구현하면서 중복으로 인한 문제가 발생할 수도 있다는 생각이 들긴 합니다🥲

Hevton replied:
안녕하세요! 이 부분에 대해서 수정을 시도해봤는데, 지금처럼 확장함수로 유효성 검증을 진행하기 위해서는 클래스 내부에 존재할 수는 없기 때문에 일반 함수로 바꿔서 클래스 내부에 구현하라는 말씀이 맞으실까요?

Gyuil-Hwnag commented on src/main/kotlin/domain/NumberGenerator.kt:
```diff
@@ -0,0 +1,5 @@
+package domain
+
+interface NumberGenerator {
+    fun generate(): Int
+}
```
인터페이스를 활용한 기능 구현 정말 좋은 것 같습니다👍

Gyuil-Hwnag replied:
다음 설정 옵션을 활성화하여 EOF 경고를 없애주세요🥲
File > Preferences > Editor > General > Ensure every saved file ends with a line break

Gyuil-Hwnag commented on src/main/kotlin/domain/TryCount.kt:
```diff
@@ -0,0 +1,20 @@
+    }
+}
+
+fun String.isValidCount(): Int {
+    val count = this.toIntOrNull()
+    requireNotNull(count) { EXCEPTION_TRY_COUNT_IS_NOT_INT }
+    require(count > 0) { EXCEPTION_INVALID_TRY_COUNT_RANGE }
+    return count
+}
+
+const val EXCEPTION_TRY_COUNT_IS_NOT_INT = "시도 횟수는 정수여야 합니다"
+const val EXCEPTION_INVALID_TRY_COUNT_RANGE = "시도 횟수는 0보다 큰 수여야 합니다"
```
해당 내용도 위의 Cars 처럼 TryCount 에서만 사용하는 함수라면 TryCount 내부에 존재하도록 구현하는 것도 고민해볼 필요가 있다고 생각이 듭니다!!🥲

Gyuil-Hwnag commented on src/main/kotlin/view/InputView.kt:
```diff
@@ -0,0 +1,20 @@
+
+class InputView {
+
+    fun readCars(): String {
+        println(HEADER_READ_CARS)
+        return readlnOrNull() ?: throw IllegalArgumentException(INPUT_IS_NULL)
+    }
+
+    fun readTryCount(): String {
+        println(HEADER_READ_TRY_COUNT)
+        return readlnOrNull() ?: throw IllegalArgumentException(INPUT_IS_NULL)
+    }
+
+    companion object {
+        const val HEADER_READ_CARS = "경주할 자동차 이름을 입력하세요.(이름은 쉼표(,) 기준으로 구분)"
+        const val INPUT_IS_NULL = "입력이 있어야 합니다"
+        const val HEADER_READ_TRY_COUNT = "시도할 횟수는 몇 회인가요?"
+    }
```
InputView에 대해서 기능 분리 너무 잘해주셨네요!!👍

Gyuil-Hwnag replied:
Input 에서만 사용하는 상수값 이라면 외부에서의 접근을 막는것도 고민해봐도 좋을 것 같습니다🤔

Gyuil-Hwnag commented on src/main/kotlin/view/OutputView.kt:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    fun printWinner(cars: Cars) {
+        val maxPosition = cars.cars.maxOf {
+            it.position
+        }
+
+        val winners = cars.cars.filter {
+            it.position == maxPosition
+        }
+
+        println("$HEADER_WINNERS " +
+                "${winners.joinToString(",") { it.name }} "
+        )
```
String Extenions 를 활용하여 깔끔하게 구현해주셨네요👍

추가로 리뷰에 남겨주신 글 보고 생각난 내용 코멘트 남겨두었습니다~
MVC로 구현한다고 했을 때 승자를 구하는 로직이 View에 존재하는게 맞는가? 라는 고민이 들긴 해서 Cars 내부에 함수로 기능을 빼두는 것도 좋을 것 같다는 생각이 듭니다!!

Hevton replied:
감사합니다 저도 그게 맞다고 느꼈습니다!

Gyuil-Hwnag commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -0,0 +1,48 @@
+
+class CarTest {
+
+    @ParameterizedTest
+    @ValueSource(strings = ["", " ", "\n", "    "])
+    fun `자동차 이름은 공백을 허용하지 않는다`(input: String) {
+        assertThatThrownBy {
+            Car(input, RandomNumberGenerator)
+        }.isExactlyInstanceOf(IllegalArgumentException::class.java)
+            .hasMessage("자동차 이름은 공백을 허용하지 않습니다")
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = ["shimji", "pangtae", "abcdefggeg"])
+    fun `자동차 이름은 5자 이하여야 한다`(input: String) {
+        assertThatThrownBy {
+            Car(input, RandomNumberGenerator)
+        }.isExactlyInstanceOf(IllegalArgumentException::class.java)
+            .hasMessage("자동차 이름은 다섯글자 이하여야 합니다")
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = [0, 1, 2, 3])
+    fun `0 부터 3 사이이면 자동차가 움직이지 않는다`(input: Int) {
+        // given
+        val car = Car("shim", ExplicitNumberGenerator(input))
+        // when
+        car.moveIfPossible()
+        assertThat(car.position == 0)
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = [4, 5, 6, 7, 8, 9])
+    fun `4 이상 9 이하이면 자동차가 움직인다`(input: Int) {
+        // given
+        val car = Car("shim", ExplicitNumberGenerator(input))
+        // when
+        car.moveIfPossible()
+        assertThat(car.position == 1)
+    }
```
테스트 코드 정말 깔끔하게 너무 잘 작성해주셨네요💯
어떤 기능을 하며 어떤 결과가 나오는지 함수명만 봐도 유추가 가능하네요👍

Gyuil-Hwnag replied:
💯💯💯
테스트함수명을 보고 어떤 테스트를 하는지에 대해서 한눈에 알아볼 수가 있네요😆

Gyuil-Hwnag reviewed (changes requested):
안녕하세요 팡태! 
이번 미션을 함께하게 된 두루라고 합니다. 잘부탁드려요!

자동자 경주 구현 고생 많으셨습니다!
미션 구현을 너무 잘 해주셨습니다 💯

고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 
화이팅입니다💪

Gyuil-Hwnag commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -0,0 +1,24 @@
+    var position: Int = 0
+        private set
+
+    init {
+        require(name.isNotBlank()) { EXCEPTION_CAR_NAME_IS_BLANK }
+        require(name.length <= 5) { EXCEPTION_INVALID_CAR_NAME_LENGTH }
+    }
```
init{ } 을 통해서 클래스 생성시에 검증을 해주는 것 정말 좋은 것 같습니다👍

Gyuil-Hwnag commented:
> 리뷰어님 안녕하세요! 저는 안드로이드 6기 팡태(황태준)입니다! 시간 내어 코드를 읽어 주셔서 정말 감사드립니다.
> 
> 미션을 진행하면서 몇가지 여쭤보고 싶은 점이 생겼는데, 혹시 답변 남겨주실 수 있으시다면 정말 감사드리겠습니다.
> 
> 1. 클래스 분리가 적절한지 피드백을 받고 싶습니다.
>    MVC 모델 기반으로 진행하고자 했습니다. 다양한 도메인을 구현했는데, 시도 횟수를 입력받는 과정까지 도메인 클래스로 구현하여 테스트 코드를 작성하는데 이점을 누리고자 했는데, 이 부분에 대한 의견을 구하고 싶습니다. 또한 이외에도 클래스 분리가 전체적으로 합리적인지 궁금합니다.
> 2. 데이터 클래스를 이용한 경우, 클래스를 이용한 경우
>    페어와 함께 이번 미션을 진행하면서 클래스 생성에 두 가지 방법을 도입했습니다.
>    A. Car 클래스는 data 클래스로 정의하고, 유효성 검증을 init에서 진행했습니다.
>    B. 그리고 이와달리 TryCount와 Cars는 일반 클래스로 정의한 뒤에 팩토리 메서드를 통해 유효성 검증과 객체 생성을 진행했습니다.
> 
> 데이터 클래스는 DTO처럼 일반적으로 데이터를 주고받는 과정에서만 사용하고, 함수 구현이나 유효성 검증 등을 진행하지 않아야 하나 싶어 일반 클래스로 구현하고 팩토리 메서드를 통해 객체를 생성함과 동시에 유효성 검증을 진행하는 방향(B)을 더 바라봐야 할지, 아니면 A 방향도 고려할 수 있을지 궁금합니다.
> 
> 3. 유효성 검증 방법
>    이번 미션에서 도메인의 유효성 검증을 진행할 때, 확장함수를 사용하였고, 해당 확장함수는 관련성이 있는 도메인 클래스 파일 내에 넣어주었습니다. 이러한 유효성 검증 방식과 분리가 현재 제 능력의 상한선이라고 생각되기에, 이에 대한 리뷰어님의 시야와 의견이 궁금합니다.
> 
> 읽어주셔서 감사드립니다.

안녕하세요 팡태~
추가적으로 질문주신거 답변 남겨두었습니다!!

전반적으로 구현하고 기능별 분리 너무 잘해주셨던 것 같습니다!! 테스트 코드 작성과 기능 확장을 위한 도메인 구분은 정말 좋은 것 같다는 생각이 듭니다😆
개인적인 생각으로는 Domain 과 Model 에서 검증하는 내용들은 느낌이 다른 검증이 들어가야 한다고 생각이 들긴 합니다!!
Model 에서의 검증은 해당 Data Class에 대해서만 판단해도 되는 내용들(글자수 제한등등)
Domain 에서의 검증은 시스템 전체적인 비즈니스 로직에 대한 판단(우승한 사람 판별등등)

미션 진행하면서 같이 고민해보도록 해봅시다 😆
화이팅입니다!!💪

Hevton commented:
@Gyuil-Hwnag 님 안녕하세요! 소중한 시간 내어주셔서 감사드립니다. 코멘트 남겨주신 부분에 대해서 궁금한 점이 생겨서 개인적으로 알아봤는데, 이해가 쉽지 않아서 직접 질문을 남기게 되었습니다.

저는 '비즈니스 로직을 담당' == 'Domain' == 'Model' 이라는 개념으로 이해하고 있으며, 이와 달리 데이터만 담고 있는 것은 Entity라고 이해하고 있습니다.

그렇기 때문에 이번 코드 구현에서 TryCount나 Car나 Cars 모두 비즈니스 로직을 담당 == 도메인 == 모델 이라고 생각했는데
남겨주신 피드백에는 도메인과 모델을 구분하고 계신 것 같아서, 어떤 클래스가 도메인 또는 모델이라고 염두하시고 말씀하시는지 이해가 잘 되지 못했습니다 ㅜ..ㅜ 도와주실 수 있다면 정말 감사드리겠습니다..

Commit:
fix(modifier): 접근 제어자 변경

- 파일 외부에서 쓰이지 않는 확장함수와 상수는 private 적용

Commit:
docs(README): 우승자 클래스 요구사항 추가

- Winners 클래스는 우승자 리스트를 갖고 있고, Cars 클래스가 결과를 토대로 생성해준다

Commit:
feat(Winners): 경주 우승자를 담는 클래스 구현

Commit:
feat(Cars): 경주 우승자를 생성하여 리턴한다

- 이전: OutputView에서 우승자를 생성하여 출력까지
- 이후: 우승자를 담는 Winners 객체를 Cars에서 만들어서 OutputView에 전달

Commit:
fix(RacingGame): 경주 우승자를 리턴하는 로직 에러 수정

- Winners를 추가한 로직을 결과에 반영하기 위해선 RacingGame 클래스의 수정도 필요했음을 누락했었다

Gyuil-Hwnag commented:
> @Gyuil-Hwnag 님 안녕하세요! 소중한 시간 내어주셔서 감사드립니다. 코멘트 남겨주신 부분에 대해서 궁금한 점이 생겨서 개인적으로 알아봤는데, 이해가 쉽지 않아서 직접 질문을 남기게 되었습니다.
> 
> 저는 '비즈니스 로직을 담당' == 'Domain' == 'Model' 이라는 개념으로 이해하고 있으며, 이와 달리 데이터만 담고 있는 것은 Entity라고 이해하고 있습니다.
> 
> 그렇기 때문에 이번 코드 구현에서 TryCount나 Car나 Cars 모두 비즈니스 로직을 담당 == 도메인 == 모델 이라고 생각했는데 남겨주신 피드백에는 도메인과 모델을 구분하고 계신 것 같아서, 어떤 클래스가 도메인 또는 모델이라고 염두하시고 말씀하시는지 이해가 잘 되지 못했습니다 ㅜ..ㅜ 도와주실 수 있다면 정말 감사드리겠습니다..

@Hevton 
팡태 안녕하세요~ 
제가 말을 애매하게 했었던 것 같네요🥲

- Entity (Cars 와 같은 Data Class)에는 해당 Entity 에서 판단 해도 되는 내용들 (예를 들어 글자수 제한이나, 이름이 빈칸이 아닌지등등)
- Domain == 비즈니스 로직을 담당 == Model 에서는 Entity들을 받아 최종적인 결과를 도출하는 내용들(예를 들어 자동차 리스트를 받아서 우승자를 판별하는 과정등)

위와 같이 생각을 하고 말한 내용이었습니다!!

개인적인 생각으로는 A와 B 모두 좋은 방법인 것 같다는 생각이 들긴 합니다!!
> A. Car 클래스는 data 클래스로 정의하고, 유효성 검증을 init에서 진행했습니다.
- 이 경우 추후 다른 사람이 볼때에도 하나의 Car 클래스 내에서 체크하는 것들에 대해서 살펴볼 수 있다는 장점이 있다고 생각이 듭니다!
B. 그리고 이와달리 TryCount와 Cars는 일반 클래스로 정의한 뒤에 팩토리 메서드를 통해 유효성 검증과 객체 생성을 진행했습니다.
- 해당의 경우에는 추후 Car 클래스의 구현이 커지는 경우나 상속받아서 사용하는 경우 Cars 팩토리 메서드로 모두 검증이 가능하다는 점에서 장점이 있다고 생각이 듭니다!

첫 미션이고 온보딩 기간이어서 궁금증이 있으면 해결 하고 가는 것도 좋지만 고민에 대해서 너무 깊게 고민하는 것보다는
미션을 같이 수행해 나가면서 같이 하나하나 풀어나가 보는 것도 좋을 것 같습니다~💪
(이런 고민들은 언제나 환영입니다👍)

Commit:
refactor(Cars): 유효성 검사 폼 변경

- 이전: 클래스 외부에 확장함수로 정의
- 이후: 클래스 내에서만 쓰이는 유효성 검증인만큼, 클래스 내부 함수로 정의

Commit:
refactor(TryCount): 유효성 검사 폼 변경

- 이전: 클래스 외부에 확장함수로 정의
- 이후: 클래스 내에서만 쓰이는 유효성 검증인만큼, 클래스 내부 함수로 정의

Hevton commented:
@Gyuil-Hwnag 님 늦은시간까지도 리뷰 남겨주시고, 좋은 말씀들과 자세한 설명 감사드립니다.
저 또한 그동안의 과거를 통해, 처음부터 너무 다양한 고민을 하는 대신에 상황에 적절하게 한 걸음 걸음 필요한 것들에 대해서만 나아가려는 마음입니다. 그에 대해서 앞으로 더 노력할 것입니다.

이번 미션에서 말씀해주신 'OutputView에서 우승자를 구하는 것 같은 로직이 있는것' 에 대한 피드백을 받고, 저 또한 OutputView에 해당 로직이 있는 것은 아쉽다고 생각되어 Winners라는 클래스를 새로 도입해서 우승자를 구하는 로직을 따로 분리하여 푸시했습니다! 좋은 말씀 감사드립니다. 

또한 확장함수 `validateCars()`` isValidCount()` 각각이 특정 클래스에서만 쓰인다면 해당 클래스 내부로 옮기는 것에 대한 의견을 말씀해주셨습니다. 말씀해주신 만큼 이 유효성 검증 확장함수는 private로 선언한다고 해도 결국 클래스 내부에서만 쓰이기에 의미를 더 명확하게 하고자 클래스 내부로 이동했습니다. 다만 팩토리 함수에서 이 유효성 검증이 필요하기에 companion object에 정의하여 푸시했습니다.

의견 남겨주셔서 감사드리며, 많이 배웠습니다 감사드립니다.

Gyuil-Hwnag commented:
> @Gyuil-Hwnag 님 늦은시간까지도 리뷰 남겨주시고, 좋은 말씀들과 자세한 설명 감사드립니다. 저 또한 그동안의 과거를 통해, 처음부터 너무 다양한 고민을 하는 대신에 상황에 적절하게 한 걸음 걸음 필요한 것들에 대해서만 나아가려는 마음입니다. 그에 대해서 앞으로 더 노력할 것입니다.
> 
> 이번 미션에서 말씀해주신 'OutputView에서 우승자를 구하는 것 같은 로직이 있는것' 에 대한 피드백을 받고, 저 또한 OutputView에 해당 로직이 있는 것은 아쉽다고 생각되어 Winners라는 클래스를 새로 도입해서 우승자를 구하는 로직을 따로 분리하여 푸시했습니다! 좋은 말씀 감사드립니다.
> 
> 또한 확장함수 `` validateCars()`` isValidCount() `` 각각이 특정 클래스에서만 쓰인다면 해당 클래스 내부로 옮기는 것에 대한 의견을 말씀해주셨습니다. 말씀해주신 만큼 이 유효성 검증 확장함수는 private로 선언한다고 해도 결국 클래스 내부에서만 쓰이기에 의미를 더 명확하게 하고자 클래스 내부로 이동했습니다. 다만 팩토리 함수에서 이 유효성 검증이 필요하기에 companion object에 정의하여 푸시했습니다.
> 
> 의견 남겨주셔서 감사드리며, 많이 배웠습니다 감사드립니다.

@Hevton 
팡테 미션 진행을 위해서 수정이 다되면 PR 요청 해주세요~💪

Hevton commented:
@Gyuil-Hwnag 수정 해서 푸시 했습니다! 리뷰요청 드렸어요!

Gyuil-Hwnag commented on src/main/kotlin/domain/Cars.kt:
```diff
@@ -0,0 +1,37 @@
+        return Winners(winners)
+    }
+
+    companion object {
+        fun from(input: String, numberGenerator: NumberGenerator) =
+            validateForm(input)
+                .map { Car(it, numberGenerator) }
+                .run { Cars(this) }
+
+        private fun validateForm(input: String): List<String> {
+            val carNames = input.split(",")
+            require(carNames.size == carNames.toSet().size) { EXCEPTION_DUPLICATED_NAME }
+            return carNames
+        }
+    }
+}
```
companion object를 잘 활용해주셨네요~👍

Gyuil-Hwnag commented on src/main/kotlin/domain/Cars.kt:
```diff
@@ -0,0 +1,37 @@
+package domain
+
+class Cars private constructor(val cars: List<Car>) {
```
private constructor 를 활용한 내용 좋네요~👍

Gyuil-Hwnag commented on src/main/kotlin/domain/RandomNumberGenerator.kt:
```diff
@@ -0,0 +1,9 @@
+    private const val UPPER_BOUND = 10
+
+    override fun generate() = Random.nextInt(UPPER_BOUND)
+}
```
다음 설정 옵션을 활성화하여 EOF 경고를 없애주세요🥲
File > Preferences > Editor > General > Ensure every saved file ends with a line break

Gyuil-Hwnag commented on src/main/kotlin/domain/TryCount.kt:
```diff
@@ -0,0 +1,20 @@
+    }
+}
+
+private const val EXCEPTION_TRY_COUNT_IS_NOT_INT = "시도 횟수는 정수여야 합니다"
+private const val EXCEPTION_INVALID_TRY_COUNT_RANGE = "시도 횟수는 0보다 큰 수여야 합니다"
```
내용 잘 구현해주셨는데, 도움될만한 내용이 있어서 코멘트 남겨두었습니다~
String 값을 관리하는 방법은 여러가지가 있을 것 같습니다!

ex)
1. 해당 const val 의 경우 TryCount에서만 사용하는 것으로 보이는데 companion object 내부에 설정하는 것도 좋아보입니다!
2. 또한 실제 Android Resource 관리할 때처럼 StringConstant 라는 object 에 생성해두는 방법도 관리하기에 좋을 것 같다는 생각이 듭니다!

Gyuil-Hwnag commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -0,0 +1,48 @@
+        assertThat(car.position == 1)
+    }
+    
+}
```
다음 설정 옵션을 활성화하여 EOF 경고를 없애주세요🥲
File > Preferences > Editor > General > Ensure every saved file ends with a line break

Gyuil-Hwnag reviewed (approved):
안녕하세요 팡태!

해당 미션 너무 고생 많으셨습니다! 
리뷰 반영 정말 잘 해주셨네요.

해당 미션은 다음 미션 진행을 위하여 머지하도록 하겠습니다.
도움이 될만한 코멘트 달았으니 확인하여 다음 미션에서 반영해주셔도 좋을 것 같습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
다음 미션도 화이팅입니다💪

Gyuil-Hwnag commented:
추가로 미션 제출전에 다음 설정으로 문법에 안 맞는 내용은 없는지 확인해 보시는 것도 좋을것 같습니다~😆
gradlew clean ktlintCheck
